outlines/models/gemini.py

```diff
@@ -0,0 +1,121 @@
+"""Integration with Gemini's API."""
+from enum import EnumMeta
+from functools import singledispatchmethod
+from types import NoneType
+from typing import Optional, Union
+
+from pydantic import BaseModel
+from typing_extensions import _TypedDictMeta  # type: ignore
+
+from outlines.prompts import Vision
+from outlines.types import Choice, Json
+
+__all__ = ["Gemini"]
+
+
+class GeminiBase:
+    """Base class for the Gemini clients.
+
+    `GeminiBase` is responsible for preparing the arguments to Gemini's
+    `generate_contents` methods: the input (prompt and possibly image), as well
+    as the output type (only JSON).
+
+    """
+
+    @singledispatchmethod
+    def format_input(self, model_input):
+        """Generate the `messages` argument to pass to the client.
+
+        Argument
+        --------
+        model_input
+            The input passed by the user.
+
+        """
+        raise NotImplementedError(
+            f"The input type {input} is not available with Gemini. The only available types are `str` and `Vision`."
+        )
+
+    @format_input.register(str)
+    def format_str_input(self, model_input: str):
+        """Generate the `messages` argument to pass to the client when the user
+        only passes a prompt.
+
+        """
+        return {"contents": [model_input]}
+
+    @format_input.register(Vision)
+    def format_vision_input(self, model_input: Vision):
+        """Generate the `messages` argument to pass to the client when the user
+        passes a prompt and an image.
+
+        """
+        return {"contents": [model_input.prompt, model_input.image]}
+
+    @singledispatchmethod
+    def format_output_type(self, output_type):
+        if output_type.__origin__ == list:
+            if len(output_type.__args__) == 1 and isinstance(
+                output_type.__args__[0], Json
+            ):
+                return {
+                    "response_mime_type": "application/json",
+                    "response_schema": list[
+                        output_type.__args__[0].original_definition
+                    ],
+                }
+            else:
+                raise TypeError
+        else:
+            raise NotImplementedError
+
+    @format_output_type.register(NoneType)
+    def format_none_output_type(self, output_type):
+        return {}
+
+    @format_output_type.register(Json)
+    def format_json_output_type(self, output_type):
+        if issubclass(output_type.original_definition, BaseModel):
```

Comment:
We could let the `Json` object do type inference for the definition, and use a simple `match` statement here.

Reply:
+1, a good indication of this is that this part stays the same:
```
            return {
                "response_mime_type": "application/json",
                "response_schema": output_type.original_definition,
            }
```
Keeping this translation clean of any other extra logic is desirable to clearly separate concerns.